examples/end_to_end.py

```diff
@@ -1,39 +1,109 @@
+import numpy as np
 import torch
 import torch.nn as nn
-import numpy as np
-from torch_radon import Radon
-
-batch_size = 8
-n_angles = 64
-image_size = 128
-channels = 4
+import torch.nn.functional as F
+from torch.utils.data import Dataset
+from torch_radon import FanBeam, Volume2D
 
-device = torch.device('cuda')
-criterion = nn.L1Loss()
 
-# Instantiate a model for the sinogram and one for the image
-sino_model = nn.Conv2d(1, channels, 5, padding=2).to(device)
-image_model = nn.Conv2d(channels, 1, 3, padding=1).to(device)
+class TestDataset(Dataset):
+    def __init__(self):
+        super().__init__()
+        img = np.load('phantom.npy')
+        
+        imgs = [
+            img,
+            np.fliplr(img).copy(),
+            np.rot90(img, k=1).copy(),
+            np.rot90(img, k=2).copy(),
+            np.rot90(img, k=3).copy(),
+        ]
+        self.imgs = [_ for _ in imgs] + [(1.0 - _.copy()) for _ in imgs]
+    
+    def __len__(self):
+        return len(self.imgs)
 
-# create empty images
-x = torch.FloatTensor(batch_size, 1, image_size, image_size).to(device)
+    def __getitem__(self, index):
+        img = self.imgs[index]
+        x = torch.from_numpy(img).unsqueeze(0).float()
+        return x
 
-# instantiate Radon transform
-angles = np.linspace(0, np.pi, n_angles)
-radon = Radon(image_size, angles)
 
-# forward projection
-sinogram = radon.forward(x)
+class TestNet(nn.Module):
+    def __init__(self, channels=3) -> None:
+        super().__init__()
+        self.sino_model = nn.Sequential(
+            nn.Conv2d(1, channels, 5, padding=2),
+            nn.Conv2d(channels, 1, 5, padding=2)
+        )
+        self.img_model = nn.Conv2d(2, 1, 3, padding=1).to(device)
+        self.det_count = 672
+    
+    def forward_sino(self, sino):
+        print('>>> Forwarding sino net.')
+        return self.sino_model(sino)
+    
+    def forward_img(self, img1, img2):
+        print('>>> Forwarding img net.')
+        img = torch.cat([img1, img2], dim=1)
+        return self.img_model(img)
+    
+    def projection(self, img, angles=None, filter=True):
+        if angles is None:
+            angles = np.linspace(0, np.pi * 2, 360, endpoint=False)
+        
+        volume = Volume2D()
+        volume.set_size(img.shape[-2], img.shape[-1])  # [B, C, H, W]
+        radon = FanBeam(self.det_count, angles, volume=volume)
+        sino = radon.forward(img)
```

Comment:
FanBeam and Volume2D classes should not be instantiated during Module execution. They should be created once in the `__init__` function of the Module like the Conv2d and Sequential Layers.

See tess/test_module.py for an example.